x11-misc/nitrogen: resolve #784320, adopt package

etc404 opened the pull request:
@gyakovlev
Messed up previous pull request, redoing it.

Signed-off-by: Dylan Wolf <etc404@protonmail.com>

gyakovlev commented:
first commit missing summary, it's just a package name. you should be able to edit by using
`git rebase -i HEAD~2` and change `pick` to `reword` or just `r` in the window that opens, next window will allow editing old message.
then just `git push --force`

also, let's do it right, how about sending patch upstream? =)
you can use your own git patch in `${FILESDIR}` even without waiting for upstream to merge it.

gyakovlev commented:
you may want to rename file to metainfo as well, as the variable too.

`appdatadir = $(datarootdir)/metainfo` becomes
`metainfodir = $(datarootdir)/metainfo`

`appdata_DATA = nitrogen.appdata.xml` becomes
`metainfo_DATA = nitrogen.metainfo.xml`

and you will have to rename the file to `.metainfo.xml` too.

that will be a proper fix.

gyakovlev commented:
if you want local hack without patching - you can just

do something like
```bash
src_install() {
    default
    dodir /usr/share/metainfo
    mv "${D}/usr/share/appdata/${PN}.appdata.xml" "${D}/usr/share/metainfo/${PN}.metainfo.xml" || die
    rmdir "${D}/usr/share/appdata" || die
}
```


but I think patching is a better exerciser in proper package maintenance.

etc404 commented:
Just to clarify, how would the upstream patch thing be different than what I've done? Or are you suggesting patching it in the original nitrogen program itself?

etc404 commented:
From what I can tell, changing the variable names (appdatadir and appdata_DATA) stops make from installing the metainfo file in the first place, and those variables don't show up in the rest of the source code- are those variable names baked in?

gyakovlev commented:
it's automake automagic, and it's critical to learn it's basics to be a good maintainer/developer.

macros are getting expanded via convoluted chain of `m4` templates and black magic =)

so basically if you have _something**dir**_  - it will get into `am__installdirs`

```
data/Makefile.in
160:am__installdirs = "$(DESTDIR)$(desktopdir)" "$(DESTDIR)$(metainfodir)"
```

and if you have _something_**_DATA**
it will get installed into _something**dir**_

end result is: data/Makefile
```Makefile
install-metainfoDATA: $(metainfo_DATA)                                                                                                                                                                                                                    
        @$(NORMAL_INSTALL)                                                                                                                                                                                                                                
        @list='$(metainfo_DATA)'; test -n "$(metainfodir)" || list=; \                                                                                                                                                                                    
        if test -n "$$list"; then \                                                                                                                                                                                                                       
          echo " $(MKDIR_P) '$(DESTDIR)$(metainfodir)'"; \                                                                                                                                                                                                
          $(MKDIR_P) "$(DESTDIR)$(metainfodir)" || exit 1; \                                                                                                                                                                                              
        fi; \                                                                                                                                                                                                                                             
        for p in $$list; do \                                                                                                                                                                                                                             
          if test -f "$$p"; then d=; else d="$(srcdir)/"; fi; \                                                                                                                                                                                           
          echo "$$d$$p"; \                                                                                                                                                                                                                                
        done | $(am__base_list) | \                                                                                                                                                                                                                       
        while read files; do \                                                                                                                                                                                                                            
          echo " $(INSTALL_DATA) $$files '$(DESTDIR)$(metainfodir)'"; \                                                                                                                                                                                   
          $(INSTALL_DATA) $$files "$(DESTDIR)$(metainfodir)" || exit $$?; \                                                                                                                                                                               
        done 
```

you can check what happens before and after autoreconf by comparing source files and generated files.

basically flow goes like this:
`Makefile.am|configure.ac -> eautoreconf -> Makefile.in|./configure -> Makefile`

so `Makefile.am` comes first, it's processed to `Makefile.in`, which is processed by ./configure into final `Makefile`
each subdir can contain a Makefile(.am|.in)

https://www.gnu.org/software/automake/manual/html_node/Data.html
https://www.gnu.org/software/automake/manual/html_node/Standard-Directory-Variables.html


It does work for me,
here's the patch I used:
```patch
From bb069ab581c2089d780f313dbe8e69c44189d5c5 Mon Sep 17 00:00:00 2001
From: Georgy Yakovlev <gyakovlev@gentoo.org>
Date: Sun, 9 Oct 2022 22:04:34 -0700
Subject: [PATCH] data/Makefile.am: install metainfo instead of appdata

---
 data/Makefile.am | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/data/Makefile.am b/data/Makefile.am
index 7fa36c5..961b761 100644
--- a/data/Makefile.am
+++ b/data/Makefile.am
@@ -1,9 +1,9 @@
 SUBDIRS = icons
 
-appdatadir = $(datarootdir)/appdata
+metainfodir = $(datarootdir)/metainfo
 desktopdir = $(datadir)/applications
 desktop_DATA = nitrogen.desktop
-appdata_DATA = nitrogen.appdata.xml
+metainfo_DATA = nitrogen.metainfo.xml
 
 UPDATE_DESKTOP = update-desktop-database $(datadir)/applications || :
 
@@ -12,5 +12,5 @@ install-data-hook:
 uninstall-hook: 
        $(UPDATE_DESKTOP)
 
-EXTRA_DIST = icon-theme-installer nitrogen.appdata.xml
+EXTRA_DIST = icon-theme-installer nitrogen.metainfo.xml
 
-- 
2.38.0
```

# note, you have to move file manually, patch can't move it

you can use `mv data/nitrogen.{appdata,metainfo}.xml || die` in `src_prepare`, just after running `default`
it's a short form of running `mv data/nitrogen.appdata.xml data/nitrogen.metainfo.xml`

```Hack
tree ../../image
../../image
└── usr
    ├── bin
    │   └── nitrogen
    ├── lib
    │   └── debug
    │       └── usr
    │           └── bin
    │               └── nitrogen.debug
    └── share
        ├── applications
        │   └── nitrogen.desktop
        ├── doc
        │   └── nitrogen-1.6.1-r1
        │       ├── AUTHORS.bz2
        │       ├── ChangeLog.bz2
        │       ├── NEWS
        │       └── README.bz2
        ├── icons
        │   └── hicolor
        │       ├── 128x128
        │       │   └── apps
        │       │       └── nitrogen.png
        │       ├── 16x16
        │       │   ├── actions
        │       │   │   ├── wallpaper-centered.png
        │       │   │   ├── wallpaper-scaled.png
        │       │   │   ├── wallpaper-tiled.png
        │       │   │   └── wallpaper-zoomed.png
        │       │   ├── apps
        │       │   │   └── nitrogen.png
        │       │   ├── devices
        │       │   │   └── video-display.png
        │       │   └── mimetypes
        │       │       └── image-x-generic.png
        │       ├── 22x22
        │       │   └── apps
        │       │       └── nitrogen.png
        │       ├── 32x32
        │       │   └── apps
        │       │       └── nitrogen.png
        │       └── 48x48
        │           └── apps
        │               └── nitrogen.png
        ├── man
        │   └── man1
        │       └── nitrogen.1.bz2
        └── metainfo
            └── nitrogen.metainfo.xml

28 directories, 20 files
```

etc404 commented:
I found my issue, I was looking in the wrong location for the metainfo file, for some reason local compilation was installing the package to /usr/_local_/share/ instead of /usr/share/. Finishing update shortly.

Flowdalic commented on x11-misc/nitrogen/nitrogen-1.6.1-r2.ebuild:
```diff
@@ -27,8 +27,13 @@ DEPEND="
 "
 
+PATCHES=(
+	"${FILESDIR}/${P}-fix-appdata-install-location.patch"
```
It is always a good idea to leave a comment to the upstream submission of the patch:

```suggestion
	"${FILESDIR}/${P}-fix-appdata-install-location.patch" # https://github.com/l3ib/nitrogen/pull/156
```

Flowdalic commented:
I am not sure what the current state is, if you where not mentored by @gyakovlev then I would usually ask to squash the commits and afterwards potentially merge it. :)

etc404 commented:
Sounds good and will do, thank you! I'll get that knocked out when I have the chance.

etc404 commented:
I've updated the patch with the link and removed the redundant part of the patch, I'd merge as well but unfortunately I do not think that is within my control.

Flowdalic commented:
lgtm, please squash
